typedargs/metadata.py

```diff
@@ -441,27 +477,39 @@ def convert_argument(self, arg_name, arg_value):
 
         self._ensure_loaded()
 
+        type_class = self.param_type_class(arg_name)
```

Comment:
There is a lot of added complexity here in multiple places to support the difference between having a `str` type_name and `type` type_class.  

We should be able to clean it all up and unify it by doing:
- move the logic into `type_system.convert_to_type`, changing the `typename` argument to `type_or_name`
- if `type_system.convert_to_type` is passed in a string, do as before
- if `type_system.convert_to_type` is passed in a type class do the following:
       - check if `isinstance(value, type_or_name)`.  If it is an instance, do nothing
       - check if `isinstance(value, str)`.  If it is an instance, try to call `FromString` to create a typed object
       - otherwise raise a `ValidationError` since you were passed in something that was not the right type and it wasn't a string you could convert to the right type.

The reason why the first `isinstance` is necessary is because the older `convert` functions in the type module implicitly check if they got passed an instance of the desired class and just return it.  `FromString`, as the name suggests expects a string so we shouldn't have the same instance check there.

If a type doesn't support conversion from string then it should just not define a `FromString` method and passing in a string should raise a validation error saying that converting to the given type from string is not supported.
This should remove the need for most of the code changes outside of `convert_to_type`.

Reply:
to finish implementation of proposed changes I need to add mapping of builtin types to our know n types from types subpackage. Will finish on this tomorrow.

Reply:
@timburke 
what is updated:
- merged latest chnages from master: PR was merged there that allows to specify validators in docstrning
- updated a test test_custom_type_class() to check validator method
- AnnotatedMetadata.param_type() now returns type class if was specified in annotations.
- updated a logic to recognize formatter: default formatter would be applied for any of (None, 'str', 'string', 'default')
        The list was (None, 'string') before. Decided to add 'str' and 'default'.
        You mentioned in #61 (with a bit different context): "`default` should be interpreted as a placeholder for `str`". So I think it should be similar at this place too.
- added a mapping builtin types (int, str, etc) to our existing type modules as described in #73.
        But this is not closing #73, at least a test should be added.
- converting from string logic and formatting return value logic are moved to `type_system`

- I have simplifyed getting type names for types from annotations.
        We still need to get type names to support logging warning message 
        in case of type info mismatch between annotations and docstring types.
        BTW, I have a doubt if we have to support this. Anyway, it works only for builtin types and Dict, List, Tuple from typing (I have hardcoded support for that three common types, which is not seeming good for me.. .)
        Maybe we should better log a warning if the given function has a type annotations and some types specified in docstring.
        Warning message would be like "Type annotations is specified. Docstring types would be ignored."
        But in this case we would have to add some syntax to specify return value formatters without specifying a type. Currently we have a syntax like `type show-as formatter: Description`.